driver-core/src/main/com/mongodb/connection/ScramShaAuthenticator.java

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.connection;
+
+import com.mongodb.AuthenticationMechanism;
+import com.mongodb.MongoCredential;
+import com.mongodb.ServerAddress;
+import com.mongodb.internal.authentication.SaslPrep;
+import org.bson.internal.Base64;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslException;
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Random;
+
+import static com.mongodb.AuthenticationMechanism.SCRAM_SHA_1;
+import static com.mongodb.AuthenticationMechanism.SCRAM_SHA_256;
+import static com.mongodb.internal.authentication.NativeAuthenticationHelper.createAuthenticationHash;
+import static java.lang.String.format;
+
+class ScramShaAuthenticator extends SaslAuthenticator {
+    private final AuthenticationMechanism authenticationMechanism;
+    private final RandomStringGenerator randomStringGenerator;
+    private final AuthenticationHashGenerator authenticationHashGenerator;
+
+    private static final int MINIMUM_ITERATION_COUNT = 4096;
+    private static final int KEY_CACHE_SIZE = 25;
+    private static final Map<String, CachedKeys> KEY_CACHE = createKeyCache();
```

Comment:
Let's discuss whether this cache should be scoped to a single Cluster instance.  So far we've been able to avoid any mutable global state (e.g. the CodecRegistry), and I'm loathe to add any.  And all the more so as the cache contains security sensitive information and there is no way to remove it except by killing the VM.

Reply:
Surely, a cache by definition is mutable? Also its authentication mechanism dependent.

Reply:
Yes, a cache by definition is mutable, but that doesn't mean it has to be global.  In this case it could be owned by DefaultClusterableServerFactory, passed into each InternalStreamConnectionFactory, and then passed into ScramShaAuthenticator

Reply:
Lets discuss.